packages/flutter/lib/src/material/i18n/README.md

```diff
@@ -0,0 +1,105 @@
+
+
+### material_en.arb Defines all of the resource IDs
+
+All of the `.arb` files whose names do not include a regional suffix contain translations for the same set of resource IDs as `material_en.arb`.
+
+For each resource ID defined for English in material_en.arb, there is an additional resource with an '@' prefix. These '@' resources are not used by the material library at run time, they just exist to inform translators about how the value will be used.
+
+```dart
+"cancelButtonLabel": "CANCEL",
+"@cancelButtonLabel": {
+  "description": "The label for cancel buttons and menu items.",
+  "type": "text"
+},
+```
+
+
+### Values with Parameters, Plurals
+
+A few of material translations contain `$variable` tokens. The material library replaces these tokens with values at run-time. For example:
+
+```dart
+"aboutListTileTitle": "About $applicationName",
+```
+
+The value for this resource ID is retrieved with a parameterized method instead of a simple getter:
+
+```dart
+MaterialLocalizations.of(context).aboutListTileTitle(yourAppTitle)
+```
+
+Plurals are handled similarly, with a lookup method that includes a quantity parameter. For example `selectedRowCountTitle` returns a string like "1 item selected" or "no items selected".
+
+```dart
+MaterialLocalizations.of(context).selectedRowCountTitle(yourRowCount)
+```
+
+Plural translations can be provided for several quantities: 0, 1, 2, "few", "many", "other". The variations are identified by a resource ID suffix which must be one of "Zero", "One", "Two", "Few", "Many", "Other". The "Other" variation is used when none of the other quantities apply. All plural resources must include a resource with the "Other" suffix. For example the English translations ('material_en.arb') for `selectedRowCountTitle` are:
+
+```dart
+"selectedRowCountTitleZero": "No items selected",
+"selectedRowCountTitleOne": "1 item selected",
+"selectedRowCountTitleOther": "$selectedRowCount items selected",
+```
+### Generated file localizations.dart: all of the localizations as a Dart Map
+
+If you look at the comment at the top of `localizations.dart` you'll see that it was manually generated using a `dev/tools` app called `gen_localizations` roughly like this:
+
+```dart
+dev/tools/gen_localizations.dart packages/flutter/lib/src/material/i18n material
+```
+
+The gen_localizations app just combines the contents of all of the .arb files into a single Dart `Map` that has entries for each .arb file's locale. The `MaterialLocalizations` class implementation uses this Map to implement the methods that lookup localized resource values.
+
+The gen_localizations app must be run by hand after .arb files have been updated. The app's first parameter is the path to this directory, the second is the file name prefix (the file name less the locale suffix) for the .arb files in this directory.
+
```

Comment:
we could generate a private class for each localization, rather than a map. that might be more efficient. It would also allow us to avoid using string replacement at runtime, we could just use literal "$foo" string interpolation...